### PR TITLE
Skil 160

### PR DESCRIPTION
**Pull request by malloc-nbytes**
Added functionality to add either students or TA's for bulk uploading.
LMS ID for students is taken as optional.
Changed the csv/xlsx parsing to "Last name, first name", email, role, (optional) LMS ID.
Added helper functions for any current and future iterations of bulk uploading.

### CODE REVIEW

**Comment by Wrimo** [comment]
I also noticed the program can partially upload a list of users. There are four users listed and only the last one is poorly formatted, the first three will be created but the last one will not. I believe it is only supposed to upload if all the users are valid.

File: BackEndFlask/controller/Routes/Bulk_upload_routes.py [hunk 1]
```diff
@@ -4,23 +4,23 @@
...
     if not file:
         print("[UploadCsv_routes /upload POST] Unsuccessfully uploaded a .csv file! No file!")
-        createBadResponse(f"Unsuccessfully uploaded a .csv file!", "No file selected!", "students")
+        createBadResponse(f"Unsuccessfully uploaded a .csv file!", "No file selected!", "users")
         return response
     extension = os.path.splitext(file.filename)
     if(extension[1]!= ".csv" and extension[1] != ".xlsx"):
         print("[UploadCsv_routes /upload POST] Unsuccessfully uploaded a .csv file! Wrong Format")
-        createBadResponse("Unsuccessfully uploaded a .csv or .xlsx file!", "Wrong Format", "students")
+        createBadResponse("Unsuccessfully uploaded a .csv or .xlsx file!", "Wrong Format", "uesrs")
```
Comment (brianlugo1) [inline BackEndFlask/controller/Routes/Bulk_upload_routes.py:23]: Typo “uesrs”!

File: BackEndFlask/Functions/test_files/test_randAssignTeams.py [hunk 2]
```diff
@@ -56,6 +56,7 @@ def test_one_ta_ten_students(flask_app_mock):
             errorMessage = "deleteOneAdminTAStudentCourse() encountered an unexpected error!"
             assert type(deleteOneAdminTAStudentCourse(result)) is not type(""), errorMessage
         except:
+            # raise
```
Comment (brianlugo1) [inline BackEndFlask/Functions/test_files/test_randAssignTeams.py:59]: We can go ahead and delete this line!

**Review by brianlugo1 — CHANGES_REQUESTED** [review]

**Review by Wrimo — CHANGES_REQUESTED** [review]

File: BackEndFlask/Functions/addStudentsAndAssignTeams.py [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+
+    """ 
+    DESCRIPTION:
+    Gets called when an error occurs when reading the roster file or something
+    went wrong with the database transaction. When called, it will delete all
+    the new users, user_courses, and team_users that were created.
+
+    PARAMETERS:
+    new_user_ids:    List[int]: A list of user_ids that were created.
+    new_user_course: List[int]: A list of *user_ids* that were created.
```
Comment (Wrimo) [inline BackEndFlask/Functions/addStudentsAndAssignTeams.py:30]: Is this comment correct? new_user_course and new_user_ids are both lists of user_ids that were created?

File: BackEndFlask/Functions/addStudentsAndAssignTeams.py [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+        if isinstance(user_id, str):
+            __uncommit_changes(new_student_ids, new_user_course_ids, new_team_id, course_id)
+            return helper_cleanup(cleanup_arr, user_id)
+
+        user_course = get_user_course_by_user_id_and_course_id(user_id, course_id)
+        if isinstance(user_course, str):
+            __uncommit_changes(new_student_ids, new_user_course_ids, new_team_id, course_id)
+            return helper_cleanup(cleanup_arr, user_course)
+
+        # TODO: Create the user_course.
```
Comment (Wrimo) [inline BackEndFlask/Functions/addStudentsAndAssignTeams.py:277]: Comment is out of date.

File: BackEndFlask/Functions/genericImport.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+def genericcsv_to_db(user_file: str, owner_id: int, course_id: int) -> None|str:
+    """ 
+    DESCRIPTION:
+    Takes a csv file and creates users of any type (student, TA, etc.)
+    and adds them to the database.
+
+    PARAMETERS:
+    user_file: str: The path to the csv file.
+    owner_id: int:  The user_id of the owner of the course.
```
Comment (Wrimo) [inline BackEndFlask/Functions/genericImport.py:22]: I believe you could remove the owner_id parameter and instead get it from the database by finding the owner_id of the course with id course_id. Not a huge deal but would simplify calling the function and potentially calling it without matching course_id and owner_id.

File: BackEndFlask/Functions/test_files/test_genericImport.py [hunk 6]
```diff
@@ -0,0 +1,803 @@
...
+            errorMessage = "deleteOneAdminCourse() encountered an unexpected error!"
+            assert type(deleteOneAdminCourse(result)) is not type(""), errorMessage
+        except:
+            errorMessage = "deleteAllUsersUserCourses() encountered an unexpected error!"
+            assert type(deleteAllUsersUserCourses(result["course_id"])) is not type(""), errorMessage
+            errorMessage = "deleteOneAdminCourse() encountered an unexpected error!"
+            assert type(deleteOneAdminCourse(result)) is not type(""), errorMessage
+            raise
+
+# test_file_not_found()
```
Comment (Wrimo) [inline BackEndFlask/Functions/test_files/test_genericImport.py:421]: Why are there lots of commented out tests? Are they broken or deprecated? Can they be deleted?

**Review by Wrimo — CHANGES_REQUESTED** [review]
Almost ready to be merged. Just a few things mostly about comments I think should be changed first.

**Pull request closed without merging**